### PR TITLE
Fix compression in ORC writer

### PR DESCRIPTION
**Pull request by vuule**
## Description
issue #12066, #12170
There is a logic error in the ORC writer that prevents use of compressed blocks in the output file. This caused all ORC files to effectively be written without compression, causing large files in many cases.
This PR makes minimal changes to fix the logic and use compressed block when compression ratio is larger than one.
Also fixed the offset at which compressed blocks are written to avoid overwriting the data as block are compacted in the output buffer.
Verified reduction in file size for the files generated in benchmarks.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/orc/stripe_enc.cu [hunk 1]
```diff
@@ -1234,7 +1235,7 @@ __global__ void __launch_bounds__(1024)
                        ? results[ss.first_block + b].bytes_written
                        : src_len;
       uint32_t blk_size24{};
-      if (results[ss.first_block + b].status == compression_status::SUCCESS) {
+      if (src_len < dst_len) {
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:1238]: That's exactly right.
Planning to add a comment.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:1238]: We can also just invert the original condition, but this gives us additional benefits with uncompressable data.

[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/src/io/orc/stripe_enc.cu:1238]: Right. This seems like a good way to do it.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/io/orc/stripe_enc.cu:1238]: Wait, so there is no longer a check condition for `SUCCESS` compression?

[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/src/io/orc/stripe_enc.cu:1238]: It is, indirectly, in the code slightly above it:

```
  auto dst_len = (results[ss.first_block + b].status == compression_status::SUCCESS)
                       ? results[ss.first_block + b].bytes_written
                       : src_len;
```

**Review by nvdbaranec — APPROVED** [review]
Tested against spark plugin + integration tests. All passed.  Suggest adding comment on the if statement for clarity.

File: cpp/src/io/orc/stripe_enc.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/io/orc/stripe_enc.cu:1238]: We don't throw any exception if the decomp process failed?

**Comment by ttnghia** [comment]
Should this also close https://github.com/rapidsai/cudf/issues/12170?

**Comment by vuule** [comment]
> Should this also close #12170?

Most likely it will address the issue. I don't see a way to verify the fix, file is not attached.

File: cpp/src/io/orc/stripe_enc.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:1240]: Added a comment. The logic should be refactored, but I don't want to make any unnecessary changes at this point.

File: cpp/src/io/orc/stripe_enc.cu [hunk 2]
```diff
@@ -1179,8 +1179,9 @@ __global__ void __launch_bounds__(256)
   num_blocks = (ss.stream_size > 0) ? (ss.stream_size - 1) / comp_blk_size + 1 : 1;
   for (uint32_t b = t; b < num_blocks; b += 256) {
     uint32_t blk_size = min(comp_blk_size, ss.stream_size - min(b * comp_blk_size, ss.stream_size));
-    inputs[ss.first_block + b]  = {src + b * comp_blk_size, blk_size};
-    auto const dst_offset       = b * (padded_block_header_size + padded_comp_block_size);
+    inputs[ss.first_block + b] = {src + b * comp_blk_size, blk_size};
+    auto const dst_offset =
+      padded_block_header_size + b * (padded_block_header_size + padded_comp_block_size);
```
Comment (ttnghia) [inline cpp/src/io/orc/stripe_enc.cu:1184]: Was this also wrongly computed before?

Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:1184]: Sort of. We insert a 3-byte header before each block when the compact the block in the output buffer. output buffer is actually same as compressed data buffer. So without the additional `padded_block_header_size` offset we would overwrite first three bytes of the first compressed block and then copy the block 3 bytes ahead, potentially doing even more damage to the block. This is why tests failed when @nvdbaranec fixed the SUCCESS condition locally. Luckily, this one change to the location where we write compressed block fixes all issues in compaction.

Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:1184]: The code that copies the blocks: https://github.com/rapidsai/cudf/blob/d335aa3dcf6a3c55dc7dbd95b94ca8e4b409daee/cpp/src/io/orc/stripe_enc.cu#L1258-L1264

**Review by ttnghia — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Review by jolorunyomi — APPROVED** [review]

**Pull request merged**